extensions/script/ExecuteScript.h

```diff
@@ -28,17 +28,95 @@
 
 #include "ScriptEngine.h"
 #include "ScriptProcessContext.h"
+#include "utils/Enum.h"
 
 namespace org {
 namespace apache {
 namespace nifi {
 namespace minifi {
+
+#ifdef PYTHON_SUPPORT
+namespace python {
+class PythonScriptEngine;
+}
+#endif  // PYTHON_SUPPORT
+
 namespace processors {
 
+class ScriptEngineFactory {
+ public:
+  ScriptEngineFactory(core::Relationship& success, core::Relationship& failure, std::shared_ptr<core::logging::Logger> logger);
+
+  template<typename T>
+  std::shared_ptr<T> createEngine() const {
+    auto engine = std::make_shared<T>();
+
+    engine->bind("log", logger_);
+    engine->bind("REL_SUCCESS", success_);
+    engine->bind("REL_FAILURE", failure_);
+
+    return engine;
+  }
+
+ private:
+  core::Relationship& success_;
+  core::Relationship& failure_;
+  std::shared_ptr<core::logging::Logger> logger_;
+};
+
+class ScriptEngineQueue {
+ public:
+  ScriptEngineQueue(uint8_t max_engine_count, ScriptEngineFactory& engine_factory, std::shared_ptr<core::logging::Logger> logger);
+
+  template<typename T>
+  std::shared_ptr<script::ScriptEngine> getScriptEngine() {
```

Comment:
How is it enforced that the returned script engine is of type `T`, especially if it's taken out of the queue? It looks like the queue can contain any kind of script engine and this function just takes out whatever is at the end.

Reply:
Good point, there is no point of having different types of script engines in a queue. Added a class level template parameter for the engine type and similar type restriction to `createEngine` in e1583ef65af13edcd0d5b07da6eee5ea84ba3cbd

Reply:
The error of the docker integration test job is related, check it out! https://github.com/apache/nifi-minifi-cpp/runs/4761689961?check_suite_focus=true#step:4:8518

Reply:
My bad I didn't have the LUA_SUPPORT flag enabled when compiled locally, the compilation issues should be fixed in 9c96d09addaa6516382a2d171f002e14f4151099